src/subhalo_tracking.cpp

```diff
@@ -175,6 +176,156 @@ inline HBTInt GetLocalHostId(HBTInt pid, const HaloSnapshot_t &halo_snap, const
+  int BoundRanking = 0;
+  for (auto particle = Subhalo.Particles.begin(); particle != Subhalo.Particles.begin() + Subhalo.Nbound; particle++)
+  {
+    if (particle->IsTracer()) // Only use tracers.
+      particle_ids[BoundRanking++] = particle->Id;
+
+    if (BoundRanking == HBTConfig.NumTracerHostFinding) // Use up to a user-defined number of tracers.
+      break;
+  }
+
+  /* Sanity checks */
+  assert(Subhalo.Particles.size() != 0); // Do not call the function for empty subhaloes.
+  assert(BoundRanking == min(Subhalo.Particles.size(),
+                             HBTConfig.NumTracerHostFinding)); // We found all expected particles
+}
+
+/* Identify which FOF is host to the particles. If we have a value of -2, that
+ * means the particle was not found in the particle information available to the
+ * local rank. */
+bool GetTracerHosts(vector<HBTInt>::iterator particle_hosts, vector<HBTInt>::const_iterator particle_ids,
+                    const HaloSnapshot_t &halo_snap, const ParticleSnapshot_t &part_snap)
+{
+  /* Initialise vector. This will make it so the code knows when to stop looking
+   * for tracers, since orphans will have all but the first with NullParticleId */
+  fill(particle_hosts, particle_hosts + HBTConfig.NumTracerHostFinding, -2);
+
+  bool MakeDecision = true;
+
+  /* Iterate over the particle list to find hosts. */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.NumTracerHostFinding; BoundRanking++)
+  {
+    // No more tracers left (should only occur for orphans!)
+    if (particle_ids[BoundRanking] == SpecialConst::NullParticleId)
+      break;
+
+    // Get host, which can be -1
+    particle_hosts[BoundRanking] = halo_snap.ParticleHash.GetIndex(particle_ids[BoundRanking]);
+
+    /* We cannot find the particle in the current rank. We will therefore need to
+     * try to find it in other ranks (and hence delay making a host decision) */
+    if (particle_hosts[BoundRanking] == -1)
+    {
+      MakeDecision = false; // Defer making a decision until we have all info.
+      if (part_snap.GetIndex(particle_ids[BoundRanking]) == SpecialConst::NullParticleId)
+        particle_hosts[BoundRanking]--; // Turns it to -2
+    }
+  }
+
+  return MakeDecision;
+}
+
+/* Assign a host to a given subgroup based on the FoF membership of the
+ * NumTracerHostFinding most bound collisionless particles. This decision is
+ * weighted by the binding energy ranking the particles had in the previous
+ * output. */
+HBTInt DecideLocalHostId(vector<HBTInt>::const_iterator particle_hosts)
+{
+  /* To store unique host candidates, and the matching score. */
+  unordered_map<HBTInt, float> CandidateHosts;
+
+  /* Iterate over the particle list, and weight each candidate score by how
+   * bound was the particle is */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.NumTracerHostFinding; BoundRanking++)
+  {
+    // Not a valid tracer (NOTE: should we break or continue here?)
+    if (particle_hosts[BoundRanking] == -2)
```

Comment:
If everything works as intended, we should expect that condition to trigger only if we have an orphan. This is because that function version only gets called if we have decided to make a decision:

https://github.com/SWIFTSIM/HBTplus/blob/4b8d6d3e3a9efa211261660982c9743847cf3992/src/subhalo_tracking.cpp#L348

For resolved subhaloes, that means we have all the information present (i.e. `TracerHosts >= 0`). For orphans, we will only have one entry filled in, with the rest defaulting to -2. 
https://github.com/SWIFTSIM/HBTplus/blob/4b8d6d3e3a9efa211261660982c9743847cf3992/src/subhalo_tracking.cpp#L213

This condition will make it so that `-2` is not a candidate to score amongst the rest.

Reply:
Ok, thanks, that makes sense now.